syntax = "proto3";

package optic;

// Sent from web client to Python server.
message UiRequest {
    oneof type {
        InitRequest init = 1;
        UserAction user_action = 2;
    }
}

message InitRequest {

}

message UserAction {
    State state = 4;

    // Used to determine which action handler to call.
    string handler_id = 1;

    Key key = 5;

    oneof type {
        bool bool = 2;
        Click click = 3;
    }
}

message Click {}

// Sent from Python server to web client.
// Multiple ServerEvent messages may be sent in response to 1 UiRequest.
message ServerEvent {
    oneof type {
        RenderEvent render = 1;
    }
}

message RenderEvent {
    Component root_component = 1;
    State state = 2;
}

// Serialized state of application. Sent from Python server to web client.
// Web client doesn't parse this data, it just sends it back to the server.
message State {
    string data = 1;
}

// Represents an instance of a component.
message Component {
    Key key = 3;
    // Type-specific component properties.
    Type type = 1;
    repeated Component children = 2;
}

message Key {
    string key = 1;
}

message Type {
    string name = 4;
    bytes value = 5;
}
